agent/src/util/device_plugin_builder.rs

```diff
@@ -0,0 +1,202 @@
+            shared,
+            node_name: env::var("AGENT_NODE_NAME")?,
+            instance_map,
+            list_and_watch_message_sender,
+            server_ender_sender: server_ender_sender.clone(),
+            device,
+        };
+
+        serve(
+            device_plugin_service,
+            socket_path.clone(),
+            server_ender_receiver,
+        )
+        .await?;
+
+        register(
+            capability_id,
+            device_endpoint,
+            &instance_name,
+            server_ender_sender,
+        )
+        .await?;
+
+        Ok(())
+    }
+}
+
+// This starts a DevicePluginServer
+pub async fn serve(
+    device_plugin_service: DevicePluginService,
+    socket_path: String,
+    server_ender_receiver: mpsc::Receiver<()>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    info!(
+        "serve - creating a device plugin server that will listen at: {}",
+        socket_path
+    );
+    tokio::fs::create_dir_all(Path::new(&socket_path[..]).parent().unwrap())
+        .await
+        .expect("Failed to create dir at socket path");
+    let mut uds = UnixListener::bind(socket_path.clone()).expect("Failed to bind to socket path");
+    let service = DevicePluginServer::new(device_plugin_service);
+    let socket_path_to_delete = socket_path.clone();
+    task::spawn(async move {
+        Server::builder()
+            .add_service(service)
+            .serve_with_incoming_shutdown(
+                uds.incoming().map_ok(unix_stream::UnixStream),
+                shutdown_signal(server_ender_receiver),
+            )
+            .await
+            .unwrap();
+        trace!(
+            "serve - gracefully shutdown ... deleting socket {}",
+            socket_path_to_delete
+        );
+        // Socket may already be deleted in the case of the kubelet restart
+        std::fs::remove_file(socket_path_to_delete).unwrap_or(());
+    });
+
+    akri_shared::uds::unix_stream::try_connect(&socket_path).await?;
+    Ok(())
+}
+
+/// This registers DevicePlugin with the kubelet.
+/// During registration, the device plugin must send
+/// (1) name of unix socket,
+/// (2) Device-Plugin API it was built against (v1beta1),
+/// (3) resource name akri.sh/device_id.
+/// If registration request to the kubelet fails, terminates DevicePluginService.
+pub async fn register(
+    capability_id: String,
+    socket_name: String,
+    instance_name: &str,
+    mut server_ender_sender: mpsc::Sender<()>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    info!(
+        "register - entered for Instance {} and socket_name: {}",
+        capability_id, socket_name
+    );
+    let op = DevicePluginOptions {
+        pre_start_required: false,
+    };
+
+    // lttp://... is a fake uri that is unused (in service_fn) but necessary for uds connection
+    let channel = Endpoint::try_from("lttp://[::]:50051")?
```

Comment:
Would it make sense to say dummy instead of lttp so that if somebody is seeing the value in the log or somewhere else, they would know it is not a typo?

Reply:
Good point. Fixed in 4bda25b